### PR TITLE
Update the polyfill to perform the operation in the next tick

### PR DESCRIPTION
**Pull request by mohsen1**
@zloirock this is mostly from your feedback [here](https://github.com/tc39/ecma262/pull/56#issuecomment-126948359). Could you review?

### CODE REVIEW

**Comment by vkurchatkin** [comment]
@mohsen1 setImmediate is not a part of Ecmascript and is not widely available. Since your proposal is based on promises it's reasonable to use `Promise.resolve().then` instead. The downside is that it actually schedules a microtasks (vs macrotask in `setImmediate`) which means that no IO or repaint will have the chance to run before parsing.

File: README.md [hunk 1]
```diff
@@ -42,19 +42,20 @@ async ()=> {
...
 #### Transpilation / Polyfill
 
-There is no way of transpilating the actual effect of this proposal. But instead we can write a polyfill that keeps API consistent while it's using the main thread for parsing and stringifying tasks.
+There is no way of transpilating the actual effect of this proposal. But instead we can write a polyfill that keeps API consistent while it's using the main thread for parsing and stringifying tasks. We are using [`setImmediate`][set-immediate] to perfom the sync operation in the next processor tick.
 
 
 ```js
 JSON.stringifyAsync = JSON.stringifyAsync || function (object) {
-  return new Promise(resolve => resolve(JSON.stringify(object)));
+  return new Promise(resolve => setImmediate(resolve(JSON.stringify(object))));
```
Comment (vkurchatkin) [inline README.md]: Still synchronous. `setImmediate` is called with `undefined`

File: README.md [hunk 2]
```diff
@@ -42,19 +42,20 @@ async ()=> {
...
 }
 
 JSON.parseAsync = JSON.parseAsync || function (string) {
-  return new Promise(resolve => resolve(JSON.parse(string)));
+  return new Promise(resolve => setImmediate(resolve(JSON.parse(string))));
```
Comment (vkurchatkin) [inline README.md]: Ditto

**Comment by ljharb** [comment]
The only way to do this with ECMAScript alone is with `Promise.resolve(object).then(JSON.stringify)`, since `setTimeout` isn't in the language either.

I'm not sure why we're concerned about the performance of a polyfill - correctness is what matters here, and the `then` example isn't even particularly slow.

**Comment by zloirock** [comment]
`setImmediate` - [task, not microtask](https://github.com/YuzuJS/setImmediate#macrotasks-and-microtasks) and isn't a part of ECMAScript (here most interesting [this proposal](https://github.com/rwaldron/tc39-notes/blob/master/es6/2014-09/sept-25.md#510-globalasap-for-enqueuing-a-microtask), but it's only conception), but yes - you understand what I mean.

**Comment by mohsen1** [comment]
Thank for feedback guys. It's using promises now. I've learned so much from you guys by just making this proposal. :smile:

File: README.md [hunk 3]
```diff
@@ -42,17 +42,22 @@ async ()=> {
...
 #### Transpilation / Polyfill
 
-There is no way of transpilating the actual effect of this proposal. But instead we can write a polyfill that keeps API consistent while it's using the main thread for parsing and stringifying tasks.
+There is no way of transpilating the actual effect of this proposal. But instead we can write a polyfill that keeps API consistent while it's using the main thread for parsing and stringifying tasks. We are using promises to perform the sync operation in the next processor tick.
 
 
 ```js
 JSON.stringifyAsync = JSON.stringifyAsync || function (object) {
-  return new Promise(resolve => resolve(JSON.stringify(object)));
+  return new Promise(resolve => {
```
Comment (vkurchatkin) [inline README.md]: no need for `new Promise`, `then` returns a promise

Comment (mohsen1) [inline README.md]: Like this?

``` js
JSON.parseAsync = JSON.stringifyAsync || function (string) {
  return Promise.resolve().then((resolve)=> resolve(JSON.parse(string)));
}
```

Comment (ljharb) [inline README.md]: As in my comment, `Promise.resolve(object).then(JSON.stringify)` is the most concise, and doesn't require an arrow function. Similarly, `Promise.resolve(string).then(JSON.parse)`.

Comment (mohsen1) [inline README.md]: Actually we completely forgot about other arguments in the polyfill.

So it should be:

``` js
JSON.parseAsync = JSON.stringifyAsync || function () {
  return Promise.resolve.apply(null, arguments).then(JSON.stringify)
}
```

will that pass all the arguments to `stringify`?

Comment (zloirock) [inline README.md]: @mohsen1 `Promise.resolve` accepts only 1 argument. Also `.apply(null` execute `.resolve` in wrong context. 

``` js
function (...args) {
  return Promise.resolve().then(_ => JSON.stringify(...args));
}
```

Comment (ljharb) [inline README.md]: ^ right, due to the other args - if the async version supports them - https://github.com/mohsen1/async-json/pull/7/files#r36045340 is good.

**Comment by mohsen1** [comment]
This should be good now.

File: README.md [hunk 4]
```diff
@@ -42,16 +42,16 @@ async ()=> {
...
 ```js
-JSON.stringifyAsync = JSON.stringifyAsync || function (object) {
-  return new Promise(resolve => resolve(JSON.stringify(object)));
+JSON.stringifyAsync = JSON.stringifyAsync || function stringifyAsync(...args) {
+ return Promise.resolve().then(()=> JSON.parse(...args));
 }
 
-JSON.parseAsync = JSON.parseAsync || function (string) {
-  return new Promise(resolve => resolve(JSON.parse(string)));
+JSON.parseAsync = JSON.parseAsync || function parseAsync(...args) {
```
Comment (ljharb) [inline README.md]: One problem with this approach - the length property of the function needs to match the number of required arguments, consistent with the rest of the spec. In other words, you'll want the async function polyfill's length to match that of the builtin sync function.

**Comment by mohsen1** [comment]
This should be good now! 🚀

File: README.md [hunk 5]
```diff
@@ -42,16 +42,16 @@ async ()=> {
...
 
-There is no way of transpilating the actual effect of this proposal. But instead we can write a polyfill that keeps API consistent while it's using the main thread for parsing and stringifying tasks.
+There is no way of transpilating the actual effect of this proposal. But instead we can write a polyfill that keeps API consistent while it's using the main thread for parsing and stringifying tasks. We are using promises to perform the sync operation in the next processor tick.
 
 
 ```js
-JSON.stringifyAsync = JSON.stringifyAsync || function (object) {
-  return new Promise(resolve => resolve(JSON.stringify(object)));
+JSON.parseAsync = JSON.parseAsync || function parseAsync(text, reviver) {
+ return Promise.resolve().then(()=> JSON.parse());
```
Comment (ljharb) [inline README.md]: this won't parse anything - perhaps you mean `JSON.parse(text, reviver)`?

File: README.md [hunk 6]
```diff
@@ -42,16 +42,16 @@ async ()=> {
...
 }
 
-JSON.parseAsync = JSON.parseAsync || function (string) {
-  return new Promise(resolve => resolve(JSON.parse(string)));
+JSON.stringifyAsync = JSON.stringifyAsync || function stringifyAsync(value, replacer, space) {
+ return Promise.resolve().then(()=> JSON.parse(value, replacer, space));
```
Comment (ljharb) [inline README.md]: and here, I assume you mean `JSON.stringify` not `JSON.parse`?

**Comment by mohsen1** [comment]
This should be good! 😑

**Pull request merged**